### PR TITLE
MM-54640 Add API to get multiple emojis by name at once

### PR DESCRIPTION
**Pull request by hmhealey**
#### Summary
It turns out that the `getCustomEmojisByName` action in the web app just sends a bunch of requests instead of making a single request to get all of the emojis at once.

This is part of the branch I implemented a while back to test an implementation of the data manager (https://github.com/mattermost/mattermost/compare/master...hh_friday-data-layer)

#### Ticket Link
MM-54640

#### Release Note
```release-note
Added API to batch requests for custom emojis on page load
```

### CODE REVIEW

File: server/channels/api4/emoji.go
Comment (hmhealey) [inline server/channels/api4/emoji.go:226]: Strictly speaking, this is a change, but this API should have checked this setting before

File: server/channels/api4/emoji.go
Comment (hmhealey) [inline server/channels/api4/emoji.go:254]: I don't know if we usually define a max value for this type of API, but it seemed like a good idea to. I set it to 200 based on the max page size for any of those APIs

File: server/channels/testlib/helper.go
Comment (hmhealey) [inline server/channels/testlib/helper.go:64]: This isn't part of this feature, but it caused me a bunch of trouble when running API tests since I have these configured via environment variables so that I don't lose them when recreating my config.json

File: webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts
Comment (hmhealey) [inline webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts:95]: I was considering making this loop more efficient, but it seems unnecessary given we shouldn't run into this case often with how we use this API

File: webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts
Comment (lieut-data) [inline webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts:94]: I don't understand this check: can you elaborate?

**Review by lieut-data — APPROVED** [review]
Seems like as solid improvement! One question, but not blocking as I'm assuming there's a good reason :)

File: webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts
Comment (larkox) [reply webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts:95]: 0/5 about making this more efficient. I feel with sets will get a bit more readable, but I also get your point this will get called not often.

File: webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts
Comment (larkox) [inline webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts:82]: Should we handle here batches for cases we have more than 200 emojis?

**Review by larkox — APPROVED** [review]
Nothing blocking, just a few comments.

Do you think you could apply this improvement also on mobile?

File: webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts:94]: This is just a shortcut to save us from having to loop through all the names/emojis to check if any were missed since, if that was the case, the results we received would be shorter than the number of names we requested

Comment (lieut-data) [reply webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts:94]: Ah, this is really nuanced. Have we measure some performance overhead here that requires this kind of optimization at the expense of code clarity?

File: webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts:95]: Good point. I'll see if I can make that a bit easier to read

File: webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts:82]: Good catch. I'll add that

File: webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts:105]: Maybe I misunderstood what you were asking. This whole block is to keep track of which emojis we've requested which weren't returned by the server because they don't exist. Some pieces of code (mostly `getCustomEmojisInText`) can get things that look like emojis but aren't real ones, so we want to track which ones aren't actually emojis that to avoid re-requesting these ones.

Do the changes I made to address Daniel's concerns make the rest of this block's purpose clearer?

**Comment by hmhealey** [comment]
> Do you think you could apply this improvement also on mobile?

Yeah, I can make a ticket to go do that. Alternatively, we wait until this is merged and make it a help wanted ticket 😀

File: webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts:105]: a

File: webapp/channels/src/packages/mattermost-redux/src/actions/emojis.test.ts
Comment (larkox) [inline webapp/channels/src/packages/mattermost-redux/src/actions/emojis.test.ts:296]: I wonder if this is in any way better to test by calling twice the `getCustomEmojisByName`, unless the logic to omit the `nonExistentEmoji` is somewhere else. And if so, I wonder if in this function instead.

File: webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts
Comment (larkox) [inline webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts:86]: Wonder if we really want to send all the requests at once. Doesn't it slightly defeat the purpose of batching? And what about if we try to fetch 10000 emojis for whatever reason? 0/5

**Review by larkox — APPROVED** [review]
Some comments on the last changes, but again nothing blocking.

File: server/channels/api4/emoji.go
Comment (M-ZubairAhmed) [reply server/channels/api4/emoji.go:254]: i am wondering whats the scenario when the length is greater than max names

File: webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts
Comment (lieut-data) [reply webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts:105]: I think the other changes are orthogonal. I'm reading `data.length !== names.length` as a way to avoid running the block of code for performance reasons, as it's otherwise safe to run through them all, if a bit redundant since if we ask for 250 emoijs and get back 250 emojis, it seems unlikely that any were missed.

I suppose I'm just wondering if the check is "necessary", because we run the code below so often that it's worth avoiding it in the first place. Maybe there's some weird bug in the future that has the server returning 250 emojis, but they aren't the ones we asked for? Would this check obscure that fact and make it harder to debug?

I'm really only 1/5 here, so no blocking required -- just thinking out loud :)

**Review by lieut-data — APPROVED** [review]

File: server/channels/api4/emoji.go
Comment (agnivade) [inline server/channels/api4/emoji.go:252]: Apologies for the drive-by comment, but `App.GetMultipleEmojiByName` already checks for this.

Comment (hmhealey) [reply server/channels/api4/emoji.go:252]: @agnivade Huh, I thought these checks tended to be on the API layer and not the app layer. Any opposition to leaving the duplicate check in?

Comment (lieut-data) [reply server/channels/api4/emoji.go:252]: Can we remove the check in the `app` layer?

Comment (hmhealey) [reply server/channels/api4/emoji.go:252]: I can do that. Presumably those methods should never be called to begin with if custom emojis are disabled

Comment (hmhealey) [reply server/channels/api4/emoji.go:252]: Actually, I'm going to leave this as-is. Most of the other emoji methods in the `api` and `app` layers duplicate these permissions checks, and there's even some repeated checks like `app.CreateEmoji` which calls `app.UploadEmojiImage`, both of which check that setting. I don't want to risk accidentally breaking something by removing a check that's expected from somewhere else (like the bulk importer or something), and I'd rather keep the methods consistent with each other than to have one be "correct" and the others not

File: server/channels/api4/emoji.go
Comment (hmhealey) [reply server/channels/api4/emoji.go:254]: It'd be hard to set up, but I think you could set it up by having DMs for 200+ users that all have different custom status emojis set

File: webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/actions/emojis.ts:105]: It was more necessary with the original version of this loop which was uglier and less performant, but it's half a performance thing since it saved us from doing the originally-O(n^2) inner loop and half a visual organization thing since the code block contains all the code needed for determining which emojis are missing

**Review by yasserfaraazkhan — APPROVED** [review]
able to add custom emoji and do not se multiple calls to `/api/v4/emoji/name/<some name>`

image

**Pull request merged**
